res_company_code/models/ir_http.py

```diff
@@ -0,0 +1,36 @@
+# Copyright (C) 2019 - Today: GRAP (http://www.grap.coop)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models
+from odoo.http import request
+
+
+class Http(models.AbstractModel):
+    _inherit = "ir.http"
+
+    def session_info(self):
+
+        result = super().session_info()
+        user = request.env.user
+        display_switch_company_menu = (
+            user.has_group("base.group_multi_company") and len(user.company_ids) > 1
+        )
+
+        # 1. Replace company name by company complete name in the session
+        #    The values are used in the switch_company_menu widget (web module)
+        # 2. Recompute sequence. (as the widget hardcod the order by sequence). See :
+        #    https://github.com/odoo/odoo/blob/16.0/addons/web/static/src/webclient/
+        #    switch_company_menu/switch_company_menu.xml#L10
+        if display_switch_company_menu:
+            sequence = 0
+            for company_id in user.company_ids.ids:
+                company = user.company_ids.filtered(lambda x: x.id == company_id)
+                sequence += 1
```

Comment:
If you iterate using `enumerate(user.company_ids)`, you wouldn't need to assign a manual sequence. Besides, companies already have sequences. Can't those be used instead?